### ISSUE TITLE
Feedback homework week 2

### ISSUE THREAD

**Issue by remarcmij** [WaelAlnadeem/hyf-javascript3#3]
Hi Wael,

Here is my feedback on your week 2 homework.

Overall, your app works fine and the code looks good. You correctly converted the callbacks to promises.

There is still the issue with the avatar picture being stretched out on a wide-screen monitor, perhaps you can look into that again.

My comments below are about how you can further improve your already working code by making it more concise and better maintainable (the shorter the code, the less chance for errors).

**1.** You can improve your `main()` function by moving the `addEventListener()` call to inside the `.then()` method, like this:

```js
function main() {
  document.getElementById('root');
  const header = createAndAppend('div', root, { class: 'header' });
  createAndAppend('h2', header, { class: 'head', html: 'HYF Repositories' });
  const select = createAndAppend('select', header, { class: 'select' });
  const main = createAndAppend('div', root, { class: 'main' });
  createAndAppend('div', main, { class: 'details' });
  createAndAppend('div', main, { class: 'contributor' });

  fetchJSON(APIUrl)
    .then(repos => {
      repos.forEach((repo, index) => {
        createAndAppend('option', select, { html: repo.name, value: index });
      });
      select.addEventListener('change', () => {
        const index = select.value;
        result(repos[index]);
      });
      result(repos[0]);
    })
    .catch(err => {
      root.innerHTML = err.message;
    });
}
```

There is no need for a second `.then()`, you can just call `result(repos[0])` directly at the end of the first `.then()`. This also makes it unnecessary to introduce a variable `repos` outside of the `.then()` method as you had in your code.

**2.** In your `result()` function, you have a reference to `repoData.gitHubAdress`. However, there is no `gitHubAddress` property on the `repoData` object. If I click the link on your web page I get an error message (did you try that yourself?). The correct property to use is `html_url`.

**3.** You have used the `createAndAppend()` function in `main()` but your `result()` function still uses the tedious, manual method with `createElement()`, `appendChild()`, etc. I think you need to be consistent and use `createAndAppend()` throughout all of your code.

**4.** If you are using a `for` loop to go through array elements, try and use a `for...of` loop instead of a traditional (old-fashioned) loop with an numerical index value. Together with the previous point **3** above, your code can be cleaned up quite a bit, making it easier to understand and maintain:

```js
function result(repoData) {
  const detailsContainer = document.querySelector('.details');
  const contributorsContainer = document.querySelector('.contributor');

  createAndAppend('h2', contributorsContainer, { html: 'contributions' });

  createAndAppend('p', detailsContainer, {
    html: 'name : <a href = "' + repoData.html_url + '">  ' + repoData.name + ' </a><br>' + 'description : ' + repoData.description + '<br>' + 'forks : ' + repoData.forks + '<br>' + 'updated :  ' + repoData.updated_at,
    class: 'detailsText'
  });

  fetchJSON(repoData.contributors_url)
    .then(contributors => {
      for (const contributor of contributors) {
        createAndAppend('img', contributorsContainer, {
          src: contributor.avatar_url,
          class: 'Img'
        });
        createAndAppend('p', contributorsContainer, {
          class: 'contribution',
          html: contributor.contributions,
        });
        createAndAppend('h3', contributorsContainer, {
          class: 'login',
          html: contributor.login
        });
      }
    })
    .catch(err => {
      root.innerHTML = err.message;
    });
}
```

**5.** When there is a network error, the only thing your web page displays is 'Not found', on an otherwise empty page. I would suggest that you at least display the header and the (potentially empty) select box. Perhaps the status code (400 something) and add some text, like 'Network error'.

**6.** Be consistent in your use of quotes as string delimiters. Either use single quotes (my preference) or double quotes throughout all of your code.

Finally, most JavaScript developers now prefer two spaces instead of four to indent their code. If you would like to try that, in VSCode you can do that by adding these User Settings:

```
"editor.detectIndentation": false,
"editor.tabSize": 2,
```

If you now reformat your code in VSCode it will indent with two space.

In conclusion, you're doing well. Good code, minor improvements possible.
